doc/builders/trivial-builders.chapter.md

```diff
@@ -67,99 +101,178 @@ writeTextFile {
+
+### `concatText` {#trivial-builder-concatText}
+
+```nix
 # Writes contents of files to /nix/store/<store path>
+
 concatText "my-file" [ file1 file2 ]
+```
+
+### `concatScript` {#trivial-builder-concatScript}
+
+```nix
+# Writes contents of files to /nix/store/<store path> and makes it executable
 
-# Writes contents of files to /nix/store/<store path>
 concatScript "my-file" [ file1 file2 ]
 ```
 
 ## `writeShellApplication` {#trivial-builder-writeShellApplication}
 
-This can be used to easily produce a shell script that has some dependencies (`runtimeInputs`). It automatically sets the `PATH` of the script to contain all of the listed inputs, sets some sanity shellopts (`errexit`, `nounset`, `pipefail`), and checks the resulting script with [`shellcheck`](https://github.com/koalaman/shellcheck).
+This function accepts one argument as an attribute set containing the following names:
 
-For example, look at the following code:
+* `name` is the name that Nix will append to the store path in the same way that `stdenv.mkDerivation` uses its `name` attribute.
+* `text` is the contents of the script.
+* `runtimeInputs` is a list of derivations.
+  For each runtime input `/nix/store/<store path>/bin` will be added to the `PATH` of the script before executing the contents of `text`.
+* `checkPhase` is an optional script that can be run to verify the output file.
+  It is empty by default and is typically used to perform syntax checks.
+
+In addition to setting the `PATH` based on `runtimeInputs`, `writeShellApplication` also sets some sane shellopts (`errexit`, `nounset`, `pipefail`)
+and checks the resulting script with `shellcheck`.
+
+For example:
 
 ```nix
 writeShellApplication {
   name = "show-nixos-org";
-
   runtimeInputs = [ curl w3m ];
-
   text = ''
+
+    # instead of ${curl}/bin/curl we can call it directly as it available in the PATH
+    # we can do the same with w3m
+
     curl -s 'https://nixos.org' | w3m -dump -T text/html
   '';
 }
 ```
 
-Unlike with normal `writeShellScriptBin`, there is no need to manually write out `${curl}/bin/curl`, setting the PATH
-was handled by `writeShellApplication`. Moreover, the script is being checked with `shellcheck` for more strict
-validation.
-
 ## `symlinkJoin` {#trivial-builder-symlinkJoin}
 
-This can be used to put many derivations into the same directory structure. It works by creating a new derivation and adding symlinks to each of the paths listed. It expects two arguments, `name`, and `paths`. `name` is the name used in the Nix store path for the created derivation. `paths` is a list of paths that will be symlinked. These paths can be to Nix store derivations or any other subdirectory contained within.
+This function can be used to merge the directory structure of several derivations into one.
+It works by creating a new derivation and adding symbolic links to each of the paths provided.
```

Comment:
Needs a bit of work, it's specifically files of the paths, but cutting off early when there's only one path for a directory.